src/emqx_auth_ldap.erl

```diff
@@ -62,14 +64,67 @@ check(ClientInfo = #{username := Username, password := Password}, AuthResult,
             {stop, AuthResult#{auth_result => ResultCode, anonymous => false}}
     end.
 
-lookup_user(Username, #{username_attr := UidAttr,
-                        match_objectclass := ObjectClass,
-                        device_dn := DeviceDn}) ->
-    Filter = eldap2:equalityMatch("objectClass", ObjectClass),
-    case search(lists:concat([UidAttr,"=", binary_to_list(Username), ",", DeviceDn]), Filter) of
-        {error, noSuchObject} ->
+lookup_user(Username, Password, #{username_attr := UidAttr,
+                                  match_objectclass := ObjectClass,
+                                  device_dn := DeviceDn,
+                                  bind_as_user := BindAsUserRequired,
+                                  custom_base_dn := CustomBaseDN} = Config) ->
+
+    Filters = maps:get(filters, Config, []),
+
+    ReplaceRules = [{"${username_attr}", UidAttr},
+                    {"${user}", binary_to_list(Username)},
+                    {"${device_dn}", DeviceDn}],
+    PasswordString = binary_to_list(Password),
+
+    %% Here the original filter could be appended to the list of filters, like
+    %% ["and",{"objectClass", ObjectClass}]
+    %% ==> "|(&((objectClass=Class)(uiAttr=someAttr))(someKey=someValue))"
+
+    SubFilters =
+        lists:map(fun({K, V}) ->
+                          {replace_vars(K, ReplaceRules), replace_vars(V, ReplaceRules)};
+                     (Op) ->
+                          Op
+                  end, Filters),
+
+    Filter =
+        case SubFilters of
+            [] -> eldap2:equalityMatch("objectClass", ObjectClass);
+            _List -> compile_filters(SubFilters, [])
+        end,
+
+    %% auth.ldap.custom_base_dn = "${username_attr}=${user},${device_dn}"
+    BaseDN = replace_vars(CustomBaseDN, ReplaceRules),
+
+    case {search(BaseDN, Filter), BindAsUserRequired} of
+        {{error, noSuchObject}, _} ->
             undefined;
-        {ok, #eldap_search_result{entries = [Entry]}} ->
+        {{ok, #eldap_search_result{entries = [Entry]}}, true} ->
+            Attributes = Entry#eldap_entry.attributes,
+            case get_value("isEnabled", Attributes) of
+                undefined ->
+                    case emqx_auth_ldap_cli:post_bind(Entry#eldap_entry.object_name, PasswordString) of
+                        ok ->
+                            Attributes;
+                        {error, Reason} ->
+                            {error, Reason}
+                    end;
+                [Val] ->
+                    case list_to_atom(string:to_lower(Val)) of
```

Comment:
would be nice to replace it to `list_to_existing_atom/1` at some point

Reply:
I know that it was inspired by the previous implementation. Just a note, I think it would be great to catch an error  if a value can't be casted to a boolean.

Reply:
Agreed. The value that can be passed is bound by the LDAP schema:
```
attributetype ( 1.3.6.1.4.1.11.2.53.2.2.3.1.2.3.1.3 NAME 'isEnabled'
	EQUALITY booleanMatch
```
I will leave it as is for now, as it is the old behaviour.